File: apps/ruby-sdk/lib/firecrawl/models/search_options.rb
# frozen_string_literal: true

module Firecrawl
  module Models
    # Options for a web search request.
    class SearchOptions
      FIELDS = %i[
        sources categories include_domains exclude_domains limit tbs location
        ignore_invalid_urls timeout scrape_options integration enterprise
      ].freeze

      attr_reader(*FIELDS)

      def initialize(**kwargs)
        FIELDS.each { |f| instance_variable_set(:"@#{f}", kwargs[f]) }
      end

      def to_h
        {
          "sources" => sources,
          "categories" => categories,
          "includeDomains" => include_domains,
          "excludeDomains" => exclude_domains,
          "limit" => limit,
          "tbs" => tbs,
          "location" => location,
          "ignoreInvalidURLs" => ignore_invalid_urls,
          "timeout" => timeout,
          "scrapeOptions" => scrape_options&.to_h,
          "integration" => integration,
          # Enterprise search options. Use ["zdr"] for end-to-end Zero Data
          # Retention or ["anon"] for anonymized search. Must be enabled for your team.
          "enterprise" => enterprise,
        }.compact
      end
    end
  end
end

File: apps/dot-net-sdk/Firecrawl/Models/SearchOptions.cs
using System.Text.Json.Serialization;

namespace Firecrawl.Models;

/// <summary>
/// Configuration options for web search.
/// </summary>
public class SearchOptions
{
    [JsonPropertyName("sources")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public List<object>? Sources { get; set; }

    [JsonPropertyName("categories")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public List<object>? Categories { get; set; }

    [JsonPropertyName("includeDomains")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public List<string>? IncludeDomains { get; set; }

    [JsonPropertyName("excludeDomains")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public List<string>? ExcludeDomains { get; set; }

    [JsonPropertyName("limit")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public int? Limit { get; set; }

    [JsonPropertyName("tbs")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Tbs { get; set; }

    [JsonPropertyName("location")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Location { get; set; }

    [JsonPropertyName("ignoreInvalidURLs")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public bool? IgnoreInvalidURLs { get; set; }

    [JsonPropertyName("timeout")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public int? Timeout { get; set; }

    [JsonPropertyName("scrapeOptions")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public ScrapeOptions? ScrapeOptions { get; set; }

    [JsonPropertyName("integration")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Integration { get; set; }

    [JsonPropertyName("country")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Country { get; set; }

    /// <summary>
    /// Enterprise search options. Use <c>["zdr"]</c> for end-to-end Zero Data
    /// Retention or <c>["anon"]</c> for anonymized search. Must be enabled for your team.
    /// </summary>
    [JsonPropertyName("enterprise")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public List<string>? Enterprise { get; set; }
}
